portal-service/src/com/liferay/portal/kernel/portlet/DefaultFriendlyURLMapper.java

```diff
@@ -268,16 +269,22 @@ protected void buildRouteParameters(
 
 		// Populate virtual parameters for instanceable portlets
 
-		if (isPortletInstanceable()) {
-			String portletId = liferayPortletURL.getPortletId();
+		String portletInstanceKey = liferayPortletURL.getPortletId();
 
-			routeParameters.put("p_p_id", portletId);
+		if (Validator.isNotNull(portletInstanceKey)) {
+			routeParameters.put("p_p_id", portletInstanceKey);
 
-			if (Validator.isNotNull(portletId) &&
-				PortletConstants.hasInstanceId(portletId)) {
+			PortletInstance portletInstance =
+				PortletInstance.fromPortletInstanceKey(portletInstanceKey);
+
+			routeParameters.put(
+				"fullInstanceId", portletInstance.getFullInstanceId());
+
+			if (portletInstance.hasInstanceId() ||
+				portletInstance.hasUserId()) {
```

Comment:
This if clause is incorrect. It should be only if (portletInstance.hasInstanceId())